### PR TITLE
HDDS-9174. Native library loader fails when system property "native.lib.tmp.dir" is not set

### PR DESCRIPTION
**Pull request by swamirishi**
## What changes were proposed in this pull request?
Native library loader fails when system property "native.lib.tmp.dir" is not set.
It fails because the library is copied to "/" instead of the cwd.

When parent has path empty. It resorts to fs.defaultParent() which is "/" so the test case fails with no permission on "/" to create a temp file.
https://github.com/bpupadhyaya/openjdk-8/blob/45af329463a45955ea2759b89cb0ebfe40570c3f/jdk/src/share/classes/java/io/File.java#L363

```java
java.io.IOException: Read-only file system
	at java.io.UnixFileSystem.createFileExclusively(Native Method)
	at java.io.File.createTempFile(File.java:2063)
	at org.apache.hadoop.hdds.utils.NativeLibraryLoader.copyResourceFromJarToTemp(NativeLibraryLoader.java:140)
	at org.apache.hadoop.hdds.utils.NativeLibraryLoader.loadLibrary(NativeLibraryLoader.java:116)
	at org.apache.hadoop.hdds.utils.db.managed.ManagedSSTDumpTool.<clinit>(ManagedSSTDumpTool.java:39)
``` 

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-9174

## How was this patch tested?
Exisiting Unit tests

### CODE REVIEW

**Comment by kerneltime** [comment]
Can you add some more description for what is the desired behavior?

**Comment by swamirishi** [comment]
> Can you add some more description for what is the desired behavior?

Updated the description.

**Review by hemantk-12 — COMMENTED** [review]
Change looks good to me. 

The only question I've is how current tests are working without it. Do we or not have test when `NATIVE_LIB_TMP_DIR` is not set? If not, we should have that.

If tests are covered in https://github.com/apache/ozone/pull/5165, that's also fine for now.

**Review by smengcl — APPROVED** [review]
+1

This was split from #5165

**Comment by swamirishi** [comment]
> 

Yeah NATIVE_LIB_TMP_DIR is set for CI runs.
https://github.com/apache/ozone/blob/a1a9471f17ca56d0bc12912f25cf922da93f26c3/hadoop-ozone/dev-support/checks/junit.sh#L32

**Comment by swamirishi** [comment]
> Change looks good to me.
> 
> The only question I've is how current tests are working without it. Do we or not have test when `NATIVE_LIB_TMP_DIR` is not set? If not, we should have that.
Yeah NATIVE_LIB_TMP_DIR is set for CI runs.
https://github.com/apache/ozone/blob/a1a9471f17ca56d0bc12912f25cf922da93f26c3/hadoop-ozone/dev-support/checks/junit.sh#L32
> 
> If tests are covered in #5165, that's also fine for now.

File: hadoop-hdds/rocks-native/pom.xml [hunk 1]
```diff
@@ -57,6 +57,11 @@
             <artifactId>mockito-junit-jupiter</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
```
Comment (hemantk-12) [inline hadoop-hdds/rocks-native/pom.xml:60]: qq: Should this be added to `hdds-common`?

**Review by hemantk-12 — COMMENTED** [review]
Changes look good to me.

File: hadoop-hdds/rocks-native/pom.xml (hunk 1 above)
Comment (swamirishi) [reply hadoop-hdds/rocks-native/pom.xml:60]: this is required only for this module for now. We can look into this later, if other modules require this later

**Comment by symious** [comment]
Seems all comments resolved, can we merge this PR?

File: hadoop-hdds/rocks-native/src/test/java/org/apache/hadoop/hdds/utils/TestNativeLibraryLoader.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+      // nativeLibraryDirectoryLocation. But load library will fail.
+      mockedNativeLibraryLoader.when(() ->
+          NativeLibraryLoader.getResourceStream(anyString()))
+          .thenReturn(new ByteArrayInputStream(new byte[]{0, 1, 2, 3}));
+      String dummyLibraryName = "dummy_lib";
+      NativeLibraryLoader.getInstance().loadLibrary(dummyLibraryName);
+      NativeLibraryLoader.isLibraryLoaded(dummyLibraryName);
+      // Checking if the resource with random was copied to a temp file.
+      Optional<File[]> libPath = Optional.ofNullable(
+          new File(nativeLibraryDirectoryLocation.orElse(""))
```
[the remark this answers is not in the document]
Comment (smengcl) [reply hadoop-hdds/rocks-native/src/test/java/org/apache/hadoop/hdds/utils/TestNativeLibraryLoader.java:86]: @swamirishi Would you address Hemant's comment above?

[the remark this answers is not in the document]
Comment (swamirishi) [reply hadoop-hdds/rocks-native/src/test/java/org/apache/hadoop/hdds/utils/TestNativeLibraryLoader.java:86]: @smengcl I have addressed the review comment. thanks for checking this.

[the remark this answers is not in the document]
Comment (hemantk-12) [reply hadoop-hdds/rocks-native/src/test/java/org/apache/hadoop/hdds/utils/TestNativeLibraryLoader.java:86]: How is this comment addressed?

[the remark this answers is not in the document]
Comment (swamirishi) [reply hadoop-hdds/rocks-native/src/test/java/org/apache/hadoop/hdds/utils/TestNativeLibraryLoader.java:86]: Sorry I didn't see the file name. I thought you were suggesting changes to be made the NativeLibraryLoader class and not the test class. Let me look into this as well.

[the remark this answers is not in the document]
Comment (swamirishi) [reply hadoop-hdds/rocks-native/src/test/java/org/apache/hadoop/hdds/utils/TestNativeLibraryLoader.java]: done

File: hadoop-hdds/rocks-native/src/test/java/org/apache/hadoop/hdds/utils/TestNativeLibraryLoader.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+      // nativeLibraryDirectoryLocation. But load library will fail.
+      mockedNativeLibraryLoader.when(() ->
+          NativeLibraryLoader.getResourceStream(anyString()))
+          .thenReturn(new ByteArrayInputStream(new byte[]{0, 1, 2, 3}));
+      String dummyLibraryName = "dummy_lib";
+      NativeLibraryLoader.getInstance().loadLibrary(dummyLibraryName);
+      NativeLibraryLoader.isLibraryLoaded(dummyLibraryName);
+      // Checking if the resource with random was copied to a temp file.
+      File[] libPath =
+          new File(nativeLibraryDirectoryLocation == null ? "" :
```
Comment (hemantk-12) [inline hadoop-hdds/rocks-native/src/test/java/org/apache/hadoop/hdds/utils/TestNativeLibraryLoader.java:82]: It brings back to the same point if we really need to test for null. You can simply have two test cases
Case 1: empty string: `""`,
Case 2: non-empty string: `File.createTempFile("prefix", "suffix").getParentFile().getAbsolutePath()`


This is test not source. In my opinion fall back to empty string in case of null doesn't make sense in test.

Comment (swamirishi) [reply hadoop-hdds/rocks-native/src/test/java/org/apache/hadoop/hdds/utils/TestNativeLibraryLoader.java:80]: by default this will go to /tmp we don't want that. We want it to copy to the current working directory.

Comment (swamirishi) [reply hadoop-hdds/rocks-native/src/test/java/org/apache/hadoop/hdds/utils/TestNativeLibraryLoader.java:80]: We are checking if the native library file gets copied to the cwd in case of null.

Comment (hemantk-12) [reply hadoop-hdds/rocks-native/src/test/java/org/apache/hadoop/hdds/utils/TestNativeLibraryLoader.java:80]: Ah, I see `nativeLibraryDirectoryLocation` is returned by `mockedNativeLibraryLoader`.

It was my miss. I didn't see `nativeLibraryDirectoryLocation` was not used other than this place.

**Review by hemantk-12 — APPROVED** [review]
LGTM.

Please check the CI failures.
